core/runtime/common/module_repository_impl.cpp

```diff
@@ -99,7 +100,11 @@ namespace kagome::runtime {
       module_instance = node.mapped();
       pool.insert(std::move(node));
     }
-    module_instance->borrow([this, state]() { release(state); });
+    module_instance->borrow([self = weak_from_this(), state]() {
+      if (!self.expired()) {
+        self.lock()->release(state);
+      }
+    });
```

Comment:
Race condition is possible between check of expiration and locking. Compare it with variant below:
```suggestion
    module_instance->borrow([wp = weak_from_this(), state]() {
      if (auto self = wp.lock()) {
        self->release(state);
      }
    });
```

Reply:
Anyway it looks weird. Probably makes sense to check **.use_count() == 1** additionally